### PR TITLE
[#IP-274] Align `GetLimitedProfile` with opt-in/opt-out preferences

### PR DESCRIPTION
**Pull request by fabriziopapi**
In order to complete the opt-in/opt-out service model implemetation, this PR upgrade the model in the fn-services and upgrade the getLimitedProfile logic (both vanilla and POST version) using the new ServicePreferencesSettings collection.

#### List of Changes

Upgrade the io-functions-commons dependency
Fix compilation errors and unit test
Add new profile-service chain of responsability to implement the opt-in/opt-out service logic
Align GetLimitedProfile and GetLimitedProfileByPOST endpoint to new logic

#### Motivation and Context

New service model requirement (opt-in/opt-out)

#### How Has This Been Tested?

Upgraded exisitng unit tests to match new model
Add unit test to cover the new profile-service implementation

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: utils/profile-services.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    .fold(
+      l => e.left(l),
+      r => e.right(r)
+    );
+
+export const handleAll = () => (
+  profile: RetrievedProfile,
+  servicesPreferencesModel: ServicesPreferencesModel,
+  serviceId: ServiceId
+): te.TaskEither<IResponseErrorNotFound, LimitedProfile> =>
```
Comment (gquadrati) [inline utils/profile-services.ts:196]: Why the need of currying here?

File: utils/profile-services.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+export const handle = (handler: IServicePreferenceHandler) => (
+  profile: RetrievedProfile,
+  servicesPreferencesModel: ServicesPreferencesModel,
+  serviceId: ServiceId
+): t.Task<e.Either<Error, LimitedProfile>> =>
+  te
+    .fromPredicate(
+      handler.isMyReposability,
+      _ => new Error("Handler not feasable!")
```
Comment (gquadrati) [inline utils/profile-services.ts:180]: typo 
```suggestion
      _ => new Error("Handler not feasible!")
```

File: utils/profile-services.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                SELECT *
+                FROM c 
+                WHERE 
+                  c.fiscalCode = @fiscalCode 
+                  AND 
+                  c.settingsVersion = @settingsVersion
+                  AND
+                  c.serviceId = @serviceId
+              `
+          })
```
Comment (gquadrati) [inline utils/profile-services.ts:99]: This is not needed. You can simply use `servicesPreferencesModel.find` method and passing it  `documentId` and  `fiscalCode`.
You can see an implementation [here](https://github.com/pagopa/io-functions-app/blob/47808ad85bde3f7c328f12cbc925b80f06a68985/GetServicePreferences/handler.ts#L105)
Note that this query could fail  with an `IResponseErrorQuery`

File: utils/profile-services.ts [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+};
+
+export const handle = (handler: IServicePreferenceHandler) => (
+  profile: RetrievedProfile,
+  servicesPreferencesModel: ServicesPreferencesModel,
+  serviceId: ServiceId
+): t.Task<e.Either<Error, LimitedProfile>> =>
+  te
+    .fromPredicate(
+      handler.isMyReposability,
```
Comment (gquadrati) [inline utils/profile-services.ts:179]: :)

File: utils/profile-services.ts [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+): te.TaskEither<IResponseErrorNotFound, LimitedProfile> =>
+  new TaskEither(
+    a.array
+      .sequence(t.taskSeq)(
+        [
+          handle(profileWithPreferenceVersionWithModeAuto),
+          handle(profileWithPreferenceVersionWithModeManual),
+          handle(profileWithModeLegacy)
+        ].map(h => h(profile, servicesPreferencesModel, serviceId))
+      )
```
Comment (gquadrati) [inline utils/profile-services.ts:205]: :)

**Review by gquadrati — CHANGES_REQUESTED** [review]
I think it's fine, I just think it's better to use thi built in `find` method for `servicesPreferencesModel`

File: GetLimitedProfile/__tests__/handler.test.ts [hunk 6]
```diff
@@ -88,10 +93,14 @@ describe("GetLimitedProfileHandler", () => {
           const mockProfileModel = ({
             findLastVersionByModelId: jest.fn(() => taskEither.of(none))
           } as unknown) as ProfileModel;
+          const mockServicesPreferencesModel = ({
+            getQueryIterator: jest.fn(() => fromLeft({}))
+          } as unknown) as ServicesPreferencesModel;
```
Comment (AleDore) [inline GetLimitedProfile/__tests__/handler.test.ts:98]: Why not extracting single costs definitions and eventually override its implementation with `mockImplementationOnce` ?

File: GetLimitedProfile/__tests__/handler.test.ts [hunk 7]
```diff
@@ -132,10 +141,14 @@ describe("GetLimitedProfileHandler", () => {
               taskEither.of(some(retrievedProfileWithInboxDisabled))
             )
           } as unknown) as ProfileModel;
+          const mockServicesPreferencesModel = ({
+            getQueryIterator: jest.fn(() => fromLeft({}))
+          } as unknown) as ServicesPreferencesModel;
```
Comment (AleDore) [inline GetLimitedProfile/__tests__/handler.test.ts:146]: ditto

File: GetLimitedProfile/__tests__/handler.test.ts [hunk 8]
```diff
@@ -172,10 +185,14 @@ describe("GetLimitedProfileHandler", () => {
               taskEither.of(some(retrievedProfile))
             )
           } as unknown) as ProfileModel;
+          const mockServicesPreferencesModel = ({
+            getQueryIterator: jest.fn(() => fromLeft({}))
+          } as unknown) as ServicesPreferencesModel;
```
Comment (AleDore) [inline GetLimitedProfile/__tests__/handler.test.ts:190]: ditto

File: GetLimitedProfile/__tests__/handler.test.ts [hunk 9]
```diff
@@ -219,10 +236,14 @@ describe("GetLimitedProfileHandler", () => {
...
```
Comment (AleDore) [inline GetLimitedProfile/__tests__/handler.test.ts:241]: ditto

File: GetLimitedProfile/__tests__/handler.test.ts [hunk 10]
```diff
@@ -266,10 +287,14 @@ describe("GetLimitedProfileHandler", () => {
...
```
Comment (AleDore) [inline GetLimitedProfile/__tests__/handler.test.ts:292]: ditto

File: GetLimitedProfile/__tests__/handler.test.ts [hunk 11]
```diff
@@ -319,10 +344,14 @@ describe("GetLimitedProfileHandler", () => {
...
```
Comment (AleDore) [inline GetLimitedProfile/__tests__/handler.test.ts:349]: ditto :)

File: GetLimitedProfileByPOST/__tests__/handler.test.ts [hunk 12]
```diff
@@ -91,10 +96,14 @@ describe("GetLimitedProfileByPOSTHandler", () => {
           const mockProfileModel = ({
             findLastVersionByModelId: jest.fn(() => taskEither.of(none))
           } as unknown) as ProfileModel;
+          const mockServicesPreferencesModel = ({
+            getQueryIterator: jest.fn(() => fromLeft({}))
+          } as unknown) as ServicesPreferencesModel;
```
Comment (AleDore) [inline GetLimitedProfileByPOST/__tests__/handler.test.ts:101]: I'd rather extract this in a single const to be reused in each test :)

File: utils/profile-services.ts (hunk 3 above)
Comment (fabriziopapi) [reply utils/profile-services.ts]: done.

File: GetLimitedProfile/__tests__/handler.test.ts (hunk 6 above)
Comment (fabriziopapi) [reply GetLimitedProfile/__tests__/handler.test.ts:98]: done.

File: utils/profile-services.ts [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+const servicePreferenceToLimitedProfile = (
+  profile: RetrievedProfile,
+  servicePreference: RetrievedServicePreference
+): LimitedProfile =>
+  retrievedProfileToLimitedProfile(profile, servicePreference.isInboxEnabled);
+
+export const profileWithPreferenceVersionWithModeAuto: IServicePreferenceHandler = {
+  handleProfile: (profile, servicesPreferencesModel, serviceId) =>
+    findServicePreference(servicesPreferencesModel, profile, serviceId).fold(
+      _l => retrievedProfileToLimitedProfile(profile, true),
```
Comment (BurnedMarshal) [inline utils/profile-services.ts:105]: Could you add a small comment that explain why we are using `true` in `retrievedProfileToLimitedProfile(profile, true)`?

Comment (BurnedMarshal) [reply utils/profile-services.ts:105]: A query error is evaluated as notFound event, and we return `sender_allowed=true` to the service calling `GetLimitedProfile`. Can we add a test on this scenario?

File: utils/profile-services.ts [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    ),
+  isMyReposability: profile =>
+    profile.servicePreferencesSettings.version >= 0 &&
+    profile.servicePreferencesSettings.mode === ServicesPreferencesModeEnum.AUTO
+};
+
+export const profileWithPreferenceVersionWithModeManual: IServicePreferenceHandler = {
+  handleProfile: (profile, servicesPreferencesModel, serviceId) =>
+    findServicePreference(servicesPreferencesModel, profile, serviceId).fold(
+      _l => retrievedProfileToLimitedProfile(profile, false),
```
Comment (BurnedMarshal) [inline utils/profile-services.ts:117]: ditto

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: utils/profile-services.ts [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+import { ServicesPreferencesModeEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/ServicesPreferencesMode";
+import * as a from "fp-ts/lib/Array";
+import { BlockedInboxOrChannelEnum } from "@pagopa/io-functions-commons/dist/generated/definitions/BlockedInboxOrChannel";
+import { TaskEither } from "fp-ts/lib/TaskEither";
+import { NonNegativeInteger } from "italia-ts-commons/lib/numbers";
+
+const NOT_FOUND_TITLE = "ServicesPreferences not found";
+
+interface IServicePreferenceHandler {
+  readonly isMyReposability: (profile: RetrievedProfile) => boolean;
```
Comment (michaeldisaro) [inline utils/profile-services.ts:25]: @balanza suggested on my branch this:

**Comment by balanza** [comment]
Closing because we're doing this in #141 

As I took over the task, I found it be easier for me to start it over rather than solve all conflicts and requested changes arose on this branch.

**Pull request closed without merging**